### PR TITLE
TMA-685: User filters and users bricks support data_product

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 1]
```diff
@@ -240,7 +241,7 @@ def call(params)
                                              create_non_existing_user_groups: create_non_existing_user_groups)
                       end
                     when 'sync_domain_client_workspaces'
-                      domain_clients = domain.clients
+                      domain_clients = domain.clients(:all, data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Identical code found in synchronize_user_filters.rb:157, synchronize_users.rb:244 (mass = 100)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 2]
```diff
@@ -122,7 +123,7 @@ def call(params)
 
             CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               client_id = row[multiple_projects_column].to_s
-              if (goodot_id && client_id == goodot_id) || domain.clients(client_id).project_uri == project.uri
+              if (goodot_id && client_id == goodot_id) || domain.clients(client_id, data_product).project_uri == project.uri
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(client_id, data_product)' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 3]
```diff
@@ -52,6 +52,7 @@ def call(params)
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) if domain_name
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 4]
```diff
@@ -163,7 +164,7 @@ def call(params)
 
             filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
               fail "Client id cannot be empty" if client_id.blank?
-              c = domain.clients(client_id)
+              c = domain.clients(client_id, data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 5]
```diff
@@ -209,7 +210,7 @@ def call(params)
                       filtered_users = new_users.select do |u|
                         fail "Column for determining the project assignement is empty for \"#{u[:login]}\"" if u[:pid].blank?
                         client_id = u[:pid].to_s
-                        (goodot_id && client_id == goodot_id) || domain.clients(client_id).project_uri == project.uri
+                        (goodot_id && client_id == goodot_id) || domain.clients(client_id, data_product).project_uri == project.uri
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'domain.clients(client_id, data_product)' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 6]
```diff
@@ -49,6 +49,7 @@ def call(params)
           domain_name = params.organization || params.domain
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
           data_source = GoodData::Helpers::DataSource.new(params.input_source)
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 7]
```diff
@@ -249,7 +250,7 @@ def call(params)
                       res = []
                       res += new_users.group_by { |u| u[:pid] }.flat_map do |client_id, users|
                         fail "Client id cannot be empty" if client_id.blank?
-                        c = domain.clients(client_id)
+                        c = domain.clients(client_id, data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/models/client.rb [hunk 8]
```diff
@@ -33,32 +33,29 @@ def [](id, opts = {})
...
               res['clients']['paging']['next']
               res['clients']['items'].each do |i|
                 p = i['client']['project']
                 tenant = client.create(GoodData::Client, i.merge('domain' => domain))
                 tenant.project = p
                 y << tenant
               end
-              url = res['clients']['paging']['next']
-              break unless url
+              uri = res['clients']['paging']['next']
```
Comment (yenkins) [inline lib/gooddata/models/client.rb]: Calls 'res['clients']' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/client.rb]: Calls 'res['clients']['paging']' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/client.rb]: Calls 'res['clients']['paging']['next']' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/client.rb [hunk 9]
```diff
@@ -33,32 +33,29 @@ def [](id, opts = {})
...
+        uri = base_uri(domain, segment.data_product)
         if id == :all
-          data_product = segment.data_product
-          tenants_uri = GoodData::DataProduct::ONE_DATA_PRODUCT_PATH % { domain_name: domain.name, id: data_product.data_product_id }
-          tenants_uri += "/clients?segment=#{CGI.escape(segment.segment_id)}"
+          uri += "?segment=#{CGI.escape(segment.segment_id)}"
           Enumerator.new do |y|
             loop do
-              res = client.get tenants_uri
+              res = client.get uri
```
Comment (yenkins) [inline lib/gooddata/models/client.rb]: Refers to 'client' more than self (maybe move it to another class?) (FeatureEnvy)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/538/ : SUCCESS in 32s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/290/ : FAILURE in 43s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/597/ : SUCCESS in 1m 53s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/539/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/291/ : SUCCESS in 37s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/598/ : SUCCESS in 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/544/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/304/ : SUCCESS in 34s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/611/ : SUCCESS in 33s

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 10]
```diff
@@ -120,16 +121,28 @@ def call(params)
...
+            end
+
+            unless goodot_id.empty? || client.nil? || (goodot_id == client.id)
+              fail "GOODOT_CUSTOM_PROJECT_ID metadata key is provided for project \"#{project.pid}\" but doesn't match \
+                    client id assigned to the project in LCM metadata. Please resolve the conflict."
+            end
+
+            filter_value = goodot_id.empty? ? client.id : goodot_id
+
+            filepath = File.open(data_source.realize(params), 'r:UTF-8')
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'file.open(data_source.realize(params), 'r:utf-8')' 6 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'data_source.realize(params)' 6 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 11]
```diff
@@ -120,16 +121,28 @@ def call(params)
...
             goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
 
-            CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
+            client = domain.clients.find { |c| c.project_uri == project.uri }
+            if goodot_id.empty? && client.nil?
+              fail "Project \"#{project.pid}\" metadata does not contain key GOODOT_CUSTOM_PROJECT_ID neither is it mapped \
+                  to a client_id in LCM metadata. We are unable to get the values for user filters."
+            end
+
+            unless goodot_id.empty? || client.nil? || (goodot_id == client.id)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'client.id' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 12]
```diff
@@ -120,16 +121,28 @@ def call(params)
             md = project.metadata
             goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
 
-            CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
+            client = domain.clients.find { |c| c.project_uri == project.uri }
+            if goodot_id.empty? && client.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'client.nil?' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'goodot_id.empty?' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 13]
```diff
@@ -141,7 +154,7 @@ def call(params)
             end
             filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
               fail "Client id cannot be empty" if client_id.blank?
-              project = domain.clients(client_id).project
+              project = domain.clients(client_id, data_product).project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(client_id, data_product)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 14]
```diff
@@ -120,16 +121,28 @@ def call(params)
...
+
+            filter_value = goodot_id.empty? ? client.id : goodot_id
+
+            filepath = File.open(data_source.realize(params), 'r:UTF-8')
+            CSV.foreach(filepath, headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
               client_id = row[multiple_projects_column].to_s
-              if (goodot_id && client_id == goodot_id) || domain.clients(client_id).project_uri == project.uri
-                filters << row
-              end
+              filters << row if client_id == filter_value
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters << row' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 15]
```diff
@@ -120,16 +121,28 @@ def call(params)
             md = project.metadata
             goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
 
-            CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
+            client = domain.clients.find { |c| c.project_uri == project.uri }
+            if goodot_id.empty? && client.nil?
+              fail "Project \"#{project.pid}\" metadata does not contain key GOODOT_CUSTOM_PROJECT_ID neither is it mapped \
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'project.pid' 9 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 16]
```diff
@@ -120,16 +121,28 @@ def call(params)
             md = project.metadata
             goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
 
-            CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
+            client = domain.clients.find { |c| c.project_uri == project.uri }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'c' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/545/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/306/ : SUCCESS in 45s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/613/ : FAILURE in 39s

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 17]
```diff
@@ -120,16 +121,28 @@ def call(params)
             md = project.metadata
             goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
 
-            CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
+            client = domain.clients(:all, data_product).find { |c| c.project_uri == project.uri }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(:all, data_product)' 2 times (DuplicateMethodCall)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/546/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/307/ : SUCCESS in 37s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/614/ : FAILURE in 35s

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 18]
```diff
@@ -44,13 +44,8 @@
...
       end
 
-      it 'uses project_id column' do
-        expect(GoodData::UserFilterBuilder).to receive(:get_filters) do |filters, _|
-          CSV.foreach(filters, headers: true, return_headers: false) do |row|
-            expect(row['project_id']).to eq '123456789'
-          end
-          []
-        end
+      it 'uses client_id column' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Identical code found in synchronize_user_filters_spec.rb:48, synchronize_users_spec.rb:47 (mass = 64)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 19]
```diff
@@ -120,7 +121,7 @@ def call(params)
             md = project.metadata
             goodot_id = md['GOODOT_CUSTOM_PROJECT_ID'].to_s
 
-            client = domain.clients.find { |c| c.project_uri == project.uri }
+            client = domain.clients(:all, data_product).find { |c| c.project_uri == project.uri }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(:all, data_product)' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 20]
```diff
@@ -317,11 +236,88 @@ def call(params)
...
+            user_group = row[user_groups_column] || row[user_groups_column.to_sym]
+            user_group = user_group.split(',').map(&:strip) if user_group
+
+            ip_whitelist = row[ip_whitelist_column] || row[ip_whitelist_column.to_sym]
+            ip_whitelist = ip_whitelist.split(',').map(&:strip) if ip_whitelist
+
+            {
+              :first_name => row[first_name_column] || row[first_name_column.to_sym],
+              :last_name => row[last_name_column] || row[last_name_column.to_sym],
+              :login => row[login_column] || row[login_column.to_sym],
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'login_column.to_sym' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'row[login_column.to_sym]' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'row[login_column]' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 21]
```diff
@@ -317,11 +236,88 @@ def call(params)
...
+          data.map do |row|
+            params.gdc_logger.debug("Processing row: #{row}")
+
+            modes = if authentication_modes.empty?
+                      row[authentication_modes_column] || row[authentication_modes_column.to_sym] || []
+                    else
+                      authentication_modes
+                    end
+
+            modes = modes.split(',').map(&:strip).map { |x| x.to_s.upcase } unless modes.is_a? Array
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Contains iterators nested 2 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has the variable name 'x' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 22]
```diff
@@ -317,11 +236,88 @@ def call(params)
...
+              :first_name => row[first_name_column] || row[first_name_column.to_sym],
+              :last_name => row[last_name_column] || row[last_name_column.to_sym],
+              :login => row[login_column] || row[login_column.to_sym],
+              :password => row[password_column] || row[password_column.to_sym],
+              :email => row[email_column] || row[login_column] || row[email_column.to_sym] || row[login_column.to_sym],
+              :role => row[role_column] || row[role_column.to_sym],
+              :sso_provider => sso_provider || row[sso_provider_column] || row[sso_provider_column.to_sym],
+              :authentication_modes => modes,
+              :user_group => user_group,
+              :pid => multiple_projects_column.nil? ? nil : (row[multiple_projects_column] || row[multiple_projects_column.to_sym]),
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 23]
```diff
@@ -317,11 +236,88 @@ def call(params)
...
 
-          puts "Printing 10 first errors"
-          puts "========================"
+          puts 'Printing 10 first errors'
+          puts '========================'
           pp errors.take(10)
           fail 'There was an error syncing users'
         end
+
+        def load_data(params, data_source)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has approx 38 statements (TooManyStatements)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 24]
```diff
@@ -240,20 +167,12 @@ def call(params)
...
                       working_client_ids = []
                       res = []
                       res += new_users.group_by { |u| u[:pid] }.flat_map do |client_id, users|
                         fail "Client id cannot be empty" if client_id.blank?
-                        c = domain.clients(client_id)
-                        if params.segments_filter && !segments_filter.include?(c.segment_uri)
-                          puts "Client #{client_id} is outside segments_filter #{params.segments_filter}"
-                          next
-                        end
+                        c = domain.clients(client_id, data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 25]
```diff
@@ -228,7 +150,12 @@ def call(params)
                     when 'sync_multiple_projects_based_on_custom_id'
                       new_users.group_by { |u| u[:pid] }.flat_map do |client_id, users|
                         fail "Client id cannot be empty" if client_id.blank?
-                        project = domain.clients(client_id).project
+                        begin
+                          project = domain.clients(client_id, data_product).project
+                        rescue RestClient::BadRequest => e
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has the variable name 'e' (UncommunicativeVariableName)

File: lib/gooddata/models/client.rb [hunk 26]
```diff
@@ -33,32 +33,30 @@ def [](id, opts = {})
...
+        uri = base_uri(domain, data_product)
         if id == :all
-          data_product = segment.data_product
-          tenants_uri = GoodData::DataProduct::ONE_DATA_PRODUCT_PATH % { domain_name: domain.name, id: data_product.data_product_id }
-          tenants_uri += "/clients?segment=#{CGI.escape(segment.segment_id)}"
+          uri += "?segment=#{CGI.escape(segment.segment_id)}" if segment
           Enumerator.new do |y|
             loop do
-              res = client.get tenants_uri
+              res = client.get uri
```
Comment (yenkins) [inline lib/gooddata/models/client.rb]: Refers to 'client' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 27]
```diff
@@ -317,11 +236,88 @@ def call(params)
...
+          phone_column                = params.phone_column || 'phone'
+          ip_whitelist_column         = params.ip_whitelist_column || 'ip_whitelist'
+          mode                        = params.sync_mode
+
+          sso_provider = params.sso_provider
+          authentication_modes = params.authentication_modes || []
+
+          multiple_projects_column = params.multiple_projects_column
+          unless multiple_projects_column
+            client_modes = %w[sync_domain_client_workspaces sync_one_project_based_on_custom_id sync_multiple_projects_based_on_custom_id]
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Style/PercentLiteralDelimiters: `%w`-literals should be delimited by `(` and `)`.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/555/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/336/ : FAILURE in 37s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/643/ : FAILURE in 31s

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/558/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/345/ : FAILURE in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/652/ : FAILURE in 33s

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 28]
```diff
@@ -18,54 +18,57 @@
...
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Similar code found in synchronize_user_filters_spec.rb:53, synchronize_users_spec.rb:47 (mass = 32)

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 29]
```diff
@@ -18,54 +18,57 @@
...
         subject.class.call(params)
       end
 
       context 'when mode is sync_domain_client_workspaces' do
         let(:mode) { 'sync_domain_client_workspaces' }
         let(:clients) { [] }
         before do
-          allow(domain).to receive(:clients).with(no_args).and_return(clients)
+          allow(domain).to receive(:clients).with(:all, nil).and_return([organization])
+          allow(domain).to receive(:clients).with(123456789).and_return(organization)
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Style/NumericLiterals: Separate every 3 digits in the integer portion of a number with underscores(_).

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 30]
```diff
@@ -18,54 +18,57 @@
...
           allow(organization).to receive(:project).and_return(project)
+          allow(organization).to receive(:client_id).and_return(123456789)
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Style/NumericLiterals: Separate every 3 digits in the integer portion of a number with underscores(_).

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/559/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/346/ : FAILURE in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/653/ : SUCCESS in 1m 51s

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/560/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/347/ : FAILURE in 1m 56s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/654/ : SUCCESS in 28s

File: spec/unit/actions/synchronize_users_spec.rb [hunk 31]
```diff
@@ -44,8 +44,8 @@
         allow(File).to receive(:open).and_return("client_id\nclient123")
       end
 
-      it 'uses client_id column' do
-        expect(domain).to receive(:clients).with('client123')
+      it 'uses client_id column to fetch client' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Similar code found in synchronize_user_filters_spec.rb:53, synchronize_users_spec.rb:47 (mass = 32)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/561/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/349/ : FAILURE in 1m 47s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/656/ : SUCCESS in 29s

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/562/ : SUCCESS in 32s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/352/ : FAILURE in 28s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/657/ : SUCCESS in 1m 45s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/564/ : SUCCESS in 33s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/354/ : SUCCESS in 1m 56s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/659/ : SUCCESS in 1m 48s

File: spec/unit/actions/collect_segment_clients_spec.rb [hunk 32]
```diff
@@ -37,4 +40,42 @@
       expect { subject.class.call(params) }.to raise_error
     end
   end
+
+  context 'when client has project in segments' do
+    let(:segment) { double('segment') }
+    let(:project) { double('project') }
+    let(:ads_client) { double('ads_client') }
+    let(:ads_response) {
```
Comment (yenkins) [inline spec/unit/actions/collect_segment_clients_spec.rb]: Style/BlockDelimiters: Avoid using `{...}` for multi-line blocks.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/565/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/355/ : FAILURE in 2m 01s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/660/ : SUCCESS in 2m 02s

File: lib/gooddata/models/client.rb [hunk 33]
```diff
@@ -33,15 +33,13 @@ def [](id, opts = {})
...
-
         client = domain.client
         fail ArgumentError, 'No client specified' if client.nil?
+        data_product = opts[:data_product] || segment.data_product
 
         if id == :all
-          data_product = segment.data_product
-          tenants_uri = GoodData::DataProduct::ONE_DATA_PRODUCT_PATH % { domain_name: domain.name, id: data_product.data_product_id }
-          tenants_uri += "/clients?segment=#{CGI.escape(segment.segment_id)}"
+          tenants_uri = base_uri(domain, data_product)
```
Comment (yenkins) [inline lib/gooddata/models/client.rb]: Calls 'base_uri(domain, data_product)' 2 times (DuplicateMethodCall)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/568/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/358/ : FAILURE in 33s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/663/ : SUCCESS in 27s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/569/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/359/ : FAILURE in 35s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/664/ : SUCCESS in 28s

**Comment by kubamahnert** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/570/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1126/ : FAILURE in 47m 00s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/882/ : SUCCESS in 24m 32s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/571/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/360/ : SUCCESS in 36s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/665/ : SUCCESS in 1m 49s

**Comment by kubamahnert** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/572/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1127/ : SUCCESS in 46m 02s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/883/ : FAILURE in 31s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/573/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/361/ : SUCCESS in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/666/ : SUCCESS in 29s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/229/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/362/ : SUCCESS in 1m 47s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/667/ : SUCCESS in 1m 44s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1128/ : ABORTED in 18m 50s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/885/ : UNSTABLE in 7m 08s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/230/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/363/ : SUCCESS in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/668/ : SUCCESS in 1m 48s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1129/ : SUCCESS in 46m 18s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/886/ : SUCCESS in 23m 53s

**Pull request merged**